docs/DesignDocs/DebuggerSupportForMultiModuleEnclaves.md

```diff
@@ -0,0 +1,268 @@
+	{
+		const char* path;
+		uint64_t path_length;
+
+		const void* base_address;
+		uint64_t size;
+	} oe_debug_enclave_t;
+	
+  ```
+
+  The `path` and `path_length` fields tell the debugger the path to the
+  enclave's ELF binary. The elf binary is expected to contain the debugging
+  information, if any, for the enclave. `oegdb` currently does not support
+  debugging enclaves whose debugging information has been stripped and moved to
+  a separate .dbg file.
+
+  The `base-address` field tells the debugger the memory address at which the
+  enclave has been loaded. The `size` field is not that important since the
+  debugger is interested only in the size of the `.text` section which it can
+  obtain from the ELF binary. The debugger associates the memory region
+  [base-address, base-address+size) with the enclave.
+
+- `oe_result_t oe_debug_notify_enclave_terminated(oe_debug_enclave_t* enclave)`
+
+  This function is called immediately after an enclave has been terminated and
+  it's memory has been deallocated. The debugger unloads the symbols for the enclave
+  and disassociates the memory region from the enclave.
+
+The debugger gains control and responds to the above notifications in the
+following ways:
+
+- Linux
+
+  The debugger sets hidden breakpoints in the above notification functions. When
+  they are called by the SDK's host-side runtime, the breakpoints are hit and
+  the debugger gets control. The debugger then reads the `oe_debug_enclave_t`
+  parameter to handle the notification.
+
+- Windows
+ 
+  Use of breakpoints by the debugger is not recommended. Instead, the Windows
+  `RaiseException` API is used by the runtime to notify the debugger.
+  
+  `oe_debug_notify_enclave_created` raises the following exception:
+  
+  ```c
+        ULONG_PTR args[1] = {(ULONG_PTR)enclave};
+		RaiseException(
+		  OE_DEBUGRT_ENCLAVE_CREATED_EVENT,
+		  0, // dwFlags
+		  1, // number of args
+		  args);
+  ```
+  
+  whereas `oe_debug_notify_enclave_terminated` raises:
+  
+  ```c
+        ULONG_PTR args[1] = {(ULONG_PTR)enclave};
+        RaiseException(
+          OE_DEBUGRT_ENCLAVE_TERMINATED_EVENT,
+          0, // dwFlags
+          1, // number of args
+          args);
+  ```
+  
+  The Windows debuggers (Visual Studio and WinDbg) consume the
+  `OE_DEBUGRT_ENCLAVE_CREATED_EVENT` and `OE_DEBUGRT_ENCLAVE_TERMINATED_EVENT`
+  events. Upon receiving the events, the debuggers invoke the appropriate
+  infrastructure to enable debugging the module.
+
+
+Use-cases for Multi-Module Enclaves
+-----------------------------------
+
+There are a few implementations of multi-module enclaves based on OE SDK:
+
+- [SGX LKL](https://github.com/lsds/sgx-lkl)
+
+  This project runs a modified linux kernel within an enclave and supports running
+  enclaves comprised of executables and shared libraries. This project supports
+  debugging of multi-module enclaves.
+  https://github.com/lsds/sgx-lkl/tree/oe_port/tools/gdb
```

Comment:
Here and in 116 and 123, it's not clear what the purpose of the link is.  121-122 at least has some text preceding the link, but what are these?